src/core/lib/security/transport/security_handshaker.c

```diff
@@ -227,24 +227,26 @@ static grpc_error *on_handshake_next_done_locked(
     return grpc_set_tsi_error_result(
         GRPC_ERROR_CREATE_FROM_STATIC_STRING("Handshake failed"), result);
   }
-  // Send data to peer.
+  // Update handshaker result.
+  if (handshaker_result != NULL) {
```

Comment:
The reason is that `if (bytes_to_send_size > 0)` and handshake completes, the check peer is done inside `on_handshake_data_sent_to_peer`, in this case, we need to set handshaker_result to the struct of security_handshaker, so that check peer can be executed correctly.